File: 2025/jimmycho/week5/2025-07-06-Linked-List-Cycle-II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = fast = head
        slow_index = 0
        while fast and fast.next:
            slow_index += 1
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                break
        else:
            return
        start = head
        meet = slow
        while meet != start:
            meet = meet.next
            if meet == slow:
                start = start.next
        return start

File: 2025/SoyoungJung/week5/2025-07-06-Linked-List-Cycle-II.ts
class DetectCycleListNode {
  val: number;
  next: DetectCycleListNode | null;
  constructor(val?: number, next?: DetectCycleListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function detectCycle(
  head: DetectCycleListNode | null
): DetectCycleListNode | null {
  let slow = head;
  let fast = head;

  while (fast !== null && fast.next !== null) {
    slow = slow!.next;
    fast = fast.next.next;

    if (slow === fast) {
      break;
    }
  }

  if (fast === null || fast.next === null) {
    return null;
  }

  slow = head;
  while (slow !== fast) {
    slow = slow!.next;
    fast = fast!.next;
  }

  return slow;
}
